src/codeu/chat/util/Timeline.java

```diff
@@ -0,0 +1,146 @@
+    public final Runnable callback;
+
+    public Event(long time, Runnable callback) {
+      this.time = time;
+      this.callback = callback;
+    }
+
+    @Override
+    public int compareTo(Event other) {
+      return Long.compare(time, other.time);
+    }
+  }
+
+  private final BlockingQueue<Event> backlog = new PriorityBlockingQueue<>();
+  private final BlockingQueue<Runnable> todo = new LinkedBlockingQueue<>();
+
+  private boolean running = true;
+
+  // This thread is used to track the time of events and moves events from the
+  // "backlog" queue to the "todo" queue when it is time to execute. They are
+  // seperated to allow this thread to be safely interrupted.
+  private final Thread scheduler = new Thread() {
+    @Override
+    public void run() {
+      while (running) {
+        final long now = System.currentTimeMillis();
+        final Event next = backlog.poll();
+        try {
+          if (next == null) {
+            Thread.sleep(10000);
+          } else if (next.time > now) {
+            // Put it back (it's not time).
+            while (!backlog.offer(next)) {
+              // force this to go through
+            }
+            Thread.sleep(next.time - now);
+          } else {
+            while (!todo.offer(next.callback)) {
+              // force this to go through
+            }
+          }
+        } catch (InterruptedException ex) {
+          // A new event was added - need to recheck all the times.
+        }
+      }
+    }
+  };
+
+  // This thread is used to run the code that was given to the time line. This
+  // worker does not need to know anything about the time. Once an event gets to
+  // here - it is considered "on time" and will be executed.
+  private final Thread executor = new Thread() {
+    @Override
+    public void run() {
+      while (running) {
+        try {
+          todo.take().run();
+        } catch (Exception ex) {
+          // Catch all exceptions here to stop any rouge action from
+          // take down the timeline.
+        }
+      }
+    }
+  };
+
+  public Timeline() {
+    scheduler.start();
+    executor.start();
+  }
+
+  // SCHEDULE NOW
+  //
+  // Add an event to the timeline so that it will occur as soon as possible.
+  public void scheduleNow(Runnable callback) {
+    scheduleAt(System.currentTimeMillis(), callback);
+  }
+
+  // SCHEDULE IN
+  //
+  // Add an event to the timeline so that it will occur in approximately in a
+  // set amount of milliseconds.
+  public void scheduleIn(long ms, Runnable callback) {
+    scheduleAt(System.currentTimeMillis() + ms, callback);
+  }
+
+  // SCHEDULE AT
+  //
+  // Add an event to the timeline so that will occur approximately at a fixed
+  // point in time.
+  public void scheduleAt(long timeMs, Runnable callback) {
+    final Event event = new Event(timeMs, callback);
+    while (!backlog.offer(event)) {
+      // force add
+    }
+    scheduler.interrupt();  // wake it up
+  }
+
+  public void stop() {
```

Comment:
No, it is never used. I have it because I did not like having a class that would spin-up threads but never join them. In the case that someone needs to stop the timeline - I did not want to provide them with a partial implementation.